edb/server/compiler/compiler.py

```diff
@@ -549,10 +550,13 @@ def _compile_ql_query(
 
         single_stmt_mode = ctx.stmt_mode is enums.CompileStatementMode.SINGLE
 
-        can_have_implicit_fields = (
-            native_out_format and
-            single_stmt_mode
-        )
+        if not ctx.inline_shapeids:
+            can_have_implicit_fields = False
```

Comment:
Why are object ids special here and not like typeid?

Reply:
I'm not sure exactly what you have in mind. I didn't want to break backward compatibility. Should this behavior be controlled entirely by the header instead of the combination of the header, protocol, and statement mode?

Reply:
Well, there can be a case of the user does not want to inline IDs but wants typenames. Your change disables typenames and typeids if object IDs are omitted.

Reply:
oh got it!

Reply:
changed